File: mobile/ios/ZeiterfassungApp/ViewModels/TimeTrackingViewModel.swift
import Foundation

@MainActor
class TimeTrackingViewModel: ObservableObject {
    @Published var trackingStatus: TrackingStatusResponse?
    @Published var isLoading = false
    @Published var isActionLoading = false
    @Published var errorMessage: String?
    @Published var actionError: String?

    private let timeService: TimeService

    init(timeService: TimeService = TimeService()) {
        self.timeService = timeService
    }

    func loadStatus(userId: String) async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            trackingStatus = try await timeService.getTrackingStatus(userId: userId)
        } catch {
            errorMessage = NSLocalizedString("common_error", comment: "")
        }
    }

    func clockIn(userId: String) async {
        await performAction(userId: userId) { try await self.timeService.clockIn() }
    }

    func clockOut(userId: String) async {
        await performAction(userId: userId) { try await self.timeService.clockOut() }
    }

    func startBreak(userId: String) async {
        await performAction(userId: userId) { try await self.timeService.startBreak() }
    }

    func endBreak(userId: String) async {
        await performAction(userId: userId) { try await self.timeService.endBreak() }
    }

    private func performAction(userId: String, action: @escaping () async throws -> TimeEntry) async {
        isActionLoading = true
        actionError = nil
        defer { isActionLoading = false }
        do {
            _ = try await action()
            await loadStatus(userId: userId)
        } catch {
            actionError = NSLocalizedString("common_error", comment: "")
        }
    }

    func clearActionError() {
        actionError = nil
    }
}

File: mobile/android/app/src/main/kotlin/com/zeiterfassung/app/ui/viewmodel/TimeTrackingViewModel.kt
package com.zeiterfassung.app.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.zeiterfassung.app.data.model.TrackingStatusResponse
import com.zeiterfassung.app.data.repository.AuthRepository
import com.zeiterfassung.app.data.repository.TimeTrackingRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class TimeTrackingUiState(
    val isLoading: Boolean = true,
    val isActionLoading: Boolean = false,
    val status: TrackingStatusResponse? = null,
    val error: String? = null,
    val actionError: String? = null,
)

@HiltViewModel
class TimeTrackingViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val timeTrackingRepository: TimeTrackingRepository,
) : ViewModel() {
    private val _uiState = MutableStateFlow(TimeTrackingUiState())
    val uiState: StateFlow<TimeTrackingUiState> = _uiState.asStateFlow()

    init {
        loadStatus()
    }

    fun loadStatus() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            val userId = authRepository.getCurrentUserId()
            if (userId == null) {
                _uiState.value = _uiState.value.copy(isLoading = false, error = "not_authenticated")
                return@launch
            }
            timeTrackingRepository
                .getStatus(userId)
                .onSuccess { status ->
                    _uiState.value = _uiState.value.copy(isLoading = false, status = status)
                }
                .onFailure {
                    _uiState.value = _uiState.value.copy(isLoading = false, error = "load_failed")
                }
        }
    }

    fun clockIn() = performAction { timeTrackingRepository.clockIn() }

    fun clockOut() = performAction { timeTrackingRepository.clockOut() }

    fun startBreak() = performAction { timeTrackingRepository.startBreak() }

    fun endBreak() = performAction { timeTrackingRepository.endBreak() }

    private fun performAction(action: suspend () -> Result<*>) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isActionLoading = true, actionError = null)
            action()
                .onSuccess { loadStatus() }
                .onFailure {
                    _uiState.value = _uiState.value.copy(isActionLoading = false, actionError = "action_failed")
                }
        }
    }

    fun clearActionError() {
        _uiState.value = _uiState.value.copy(actionError = null)
    }
}
